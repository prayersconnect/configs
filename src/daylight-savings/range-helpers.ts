import { getDSTEnd, getDSTStart } from '.';
import { DateTime } from 'luxon';

const formatString = 'yyyy-MM-dd';

function parseDateInZone(date: Date, timezone: string): DateTime {
  date.setHours(3, 0, 0, 0);
  return DateTime.fromJSDate(date).setZone(timezone, { keepLocalTime: true });
}

function getDSTRanges(dstStart: DateTime, dstEnd: DateTime, country: string) {
  let ranges = [];
  if (dstStart.year === dstEnd.year) {
    ranges.push([
      `DST Period ${dstStart.year}`,
      `${dstStart.toFormat(formatString)}`,
      `${dstEnd.toFormat(formatString)}`,
    ]);
  } else {
    const prevYearDstEnd = parseDateInZone(
      getDSTEnd(country, dstStart.year - 1) as Date,
      dstEnd.zoneName
    ); //@todo fix when getDSTEnd returns null

    const prevYearDstStart = parseDateInZone(
      getDSTStart(country, dstStart.year - 1) as Date,
      dstStart.zoneName
    ); //@todo fix when getDSTEnd returns null

    ranges.push([
      `DST Period ${prevYearDstStart.monthShort}, ${prevYearDstStart.year} - ${prevYearDstEnd.monthShort}, ${prevYearDstEnd.year}`,
      prevYearDstStart.toFormat(formatString),
      prevYearDstEnd.toFormat(formatString),
    ]);

    ranges.push([
      `DST Period ${dstStart.monthShort}, ${dstStart.year} - ${dstEnd.monthShort}, ${dstEnd.year}`,
      dstStart.toFormat(formatString),
      dstEnd.toFormat(formatString),
    ]);
  }

  return ranges;
}

function getNonDSTRanges(
  dstStart: DateTime,
  dstEnd: DateTime,
  country: string
) {
  const prevYearDSTEndDate = getDSTEnd(country, dstStart.year - 1) as Date;

  const ranges = [];
  if (dstStart.year === dstEnd.year) {
    ranges.push([
      `Non-DST Period ${dstStart.startOf('year').monthShort}, ${
        dstStart.year
      } - ${dstStart.monthShort}, ${dstStart.year}`,
      dstStart.startOf('year').toFormat(formatString),
      dstStart.toFormat(formatString),
    ]);

    ranges.push([
      `Non-DST Period ${dstEnd.monthShort}, ${dstStart.year} - ${
        dstEnd.endOf('year').monthShort
      }, ${dstStart.year}`,
      dstEnd.toFormat(formatString),
      dstEnd.endOf('year').toFormat(formatString),
    ]);
  } else {
    if (prevYearDSTEndDate) {
      const prevYearDstEnd = parseDateInZone(
        prevYearDSTEndDate,
        dstStart.zoneName
      );

      ranges.push([
        `Non-DST Period ${prevYearDstEnd.monthShort}, ${prevYearDstEnd.year} - ${dstStart.monthShort}, ${dstStart.year}`,
        prevYearDstEnd.toFormat(formatString),
        dstStart.toFormat(formatString),
      ]);
    }
  }
  return ranges;
}

export function getRangesForYear(
  year: number,
  country: string,
  timezone: string
) {
  const today = parseDateInZone(DateTime.local().toJSDate(), timezone).set({
    year,
  });
  const dstStart = parseDateInZone(
    getDSTStart(country, today.year) as Date,
    timezone
  );

  console.log(
    'dstStart',
    getDSTStart(country, today.year),
    dstStart.toISODate()
  );
  const dstEnd = parseDateInZone(
    getDSTEnd(country, today.year) as Date,
    timezone
  );

  let ranges = [
    [
      `Year ${dstStart.year}`,
      `${dstStart.startOf('year').toFormat(formatString)}`,
      `${dstStart.endOf('year').toFormat(formatString)}`,
    ],
  ];

  return ranges
    .concat(getDSTRanges(dstStart, dstEnd, country))
    .concat(getNonDSTRanges(dstStart, dstEnd, country));
}

export function getQuickSelectorRanges(country: string, timezone: string) {
  const today = DateTime.local().setZone(timezone);

  let ranges = [['------------Select a Range------------', '']];
  ranges = ranges.concat(getRangesForYear(today.year, country, timezone));
  ranges = ranges.concat([['------------', '']]);
  ranges = ranges.concat(getRangesForYear(today.year + 1, country, timezone));
  return ranges;
}
