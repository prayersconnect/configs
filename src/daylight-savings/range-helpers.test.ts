import { getRangesForYear } from './range-helpers';

describe('quick range helpers', () => {
  describe('getRangesForYear', () => {
    it('returns correct ranges for Australia', () => {
      expect(
        getRangesForYear(2022, 'Australia', 'Australia/Sydney')
      ).toMatchSnapshot();
    });
    fit('returns correct ranges for USA', () => {
      expect(
        getRangesForYear(2022, 'United States', 'America/New_York')
      ).toMatchSnapshot();
    });
  });
});
