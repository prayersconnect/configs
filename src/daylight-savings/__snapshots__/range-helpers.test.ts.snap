// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quick range helpers getRangesForYear returns correct ranges for Australia 1`] = `
[
  [
    "Year 2022",
    "2022-01-01",
    "2022-12-31",
  ],
  [
    "DST Period Oct, 2021 - Apr, 2022",
    "2021-10-02",
    "2022-04-02",
  ],
  [
    "DST Period Oct, 2022 - Apr, 2023",
    "2022-10-01",
    "2023-04-01",
  ],
  [
    "Non-DST Period Apr, 2022 - Oct, 2022",
    "2022-04-02",
    "2022-10-01",
  ],
]
`;

exports[`quick range helpers getRangesForYear returns correct ranges for USA 1`] = `
[
  [
    "Year 2022",
    "2022-01-01",
    "2022-12-31",
  ],
  [
    "DST Period 2022",
    "2022-03-12",
    "2022-11-05",
  ],
  [
    "Non-DST Period Jan, 2022 - Mar, 2022",
    "2022-01-01",
    "2022-03-12",
  ],
  [
    "Non-DST Period Nov, 2022 - Dec, 2022",
    "2022-11-05",
    "2022-12-31",
  ],
]
`;
